// kepler.gl/dist/processors/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "processGeojson", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.processGeojson;
  }
});
Object.defineProperty(exports, "processCsvData", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.processCsvData;
  }
});
Object.defineProperty(exports, "processRowObject", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.processRowObject;
  }
});
Object.defineProperty(exports, "processKeplerglJSON", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.processKeplerglJSON;
  }
});
Object.defineProperty(exports, "analyzerTypeToFieldType", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.analyzerTypeToFieldType;
  }
});
Object.defineProperty(exports, "getFieldsFromData", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.getFieldsFromData;
  }
});
Object.defineProperty(exports, "parseCsvRowsByFieldType", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.parseCsvRowsByFieldType;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.Processors;
  }
});
Object.defineProperty(exports, "getFileHandler", {
  enumerable: true,
  get: function get() {
    return _fileHandler.getFileHandler;
  }
});
Object.defineProperty(exports, "getFileType", {
  enumerable: true,
  get: function get() {
    return _fileHandler.getFileType;
  }
});

var _dataProcessor = require("./data-processor");

var _fileHandler = require("./file-handler");
