// kepler.gl/dist/utils/file-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.processFileToLoad = void 0;

var _utils = require("./utils");

var _fileHandler = require("../processors/file-handler");

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
var processFileToLoad = function processFileToLoad(file) {
  return {
    fileBlob: file,
    info: {
      id: (0, _utils.generateHashId)(4),
      label: file.name,
      size: file.size
    },
    handler: (0, _fileHandler.getFileHandler)(file)
  };
};

exports.processFileToLoad = processFileToLoad;
