// kepler.gl/dist/components/side-panel/layer-panel/text-label-panel.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = require("../../common/styled-components");

var _icons = require("../../common/icons");

var _colorSelector = _interopRequireDefault(require("./color-selector"));

var _fieldSelector = _interopRequireDefault(require("../../common/field-selector"));

var _itemSelector = _interopRequireDefault(require("../../common/item-selector/item-selector"));

var _layerConfigGroup = _interopRequireWildcard(require("./layer-config-group"));

var _rangeSlider = _interopRequireDefault(require("../../common/range-slider"));

var _layerFactory = require("../../../layers/layer-factory");

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
var TextLabelPanel =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(TextLabelPanel, _Component);

  function TextLabelPanel() {
    (0, _classCallCheck2["default"])(this, TextLabelPanel);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(TextLabelPanel).apply(this, arguments));
  }

  (0, _createClass2["default"])(TextLabelPanel, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          updateLayerTextLabel = _this$props.updateLayerTextLabel,
          textLabel = _this$props.textLabel,
          fields = _this$props.fields;
      var currentFields = textLabel.map(function (tl) {
        return tl.field && tl.field.name;
      }).filter(function (d) {
        return d;
      });
      return _react["default"].createElement(_layerConfigGroup["default"], {
        label: 'label',
        collapsible: true
      }, _react["default"].createElement(_layerConfigGroup.ConfigGroupCollapsibleHeader, null, _react["default"].createElement(_fieldSelector["default"], {
        fields: fields,
        value: currentFields,
        onSelect: function onSelect(selected) {
          return updateLayerTextLabel('all', 'fields', selected);
        },
        multiSelect: true
      })), _react["default"].createElement(_layerConfigGroup.ConfigGroupCollapsibleContent, null, textLabel.map(function (tl, idx) {
        return _react["default"].createElement("div", {
          key: tl.field ? tl.field.name : 'null'
        }, _react["default"].createElement(_styledComponents.PanelLabel, null, "Label ".concat(idx + 1)), _react["default"].createElement(_styledComponents.SidePanelSection, null, _react["default"].createElement(_fieldSelector["default"], {
          fields: fields,
          value: tl.field && tl.field.name || 'Select a field',
          placeholder: 'empty',
          onSelect: function onSelect(v) {
            return updateLayerTextLabel(idx, 'field', v);
          },
          erasable: true
        })), _react["default"].createElement(_styledComponents.SidePanelSection, null, _react["default"].createElement(_styledComponents.PanelLabel, null, "Font size"), _react["default"].createElement(_rangeSlider["default"], (0, _extends2["default"])({}, _layerFactory.LAYER_TEXT_CONFIGS.fontSize, {
          value1: tl.size,
          isRange: false,
          onChange: function onChange(v) {
            return updateLayerTextLabel(idx, 'size', v[1]);
          }
        }))), _react["default"].createElement(_styledComponents.SidePanelSection, null, _react["default"].createElement(_styledComponents.PanelLabel, null, "Font color"), _react["default"].createElement(_colorSelector["default"], {
          colorSets: [{
            selectedColor: tl.color,
            setColor: function setColor(v) {
              return updateLayerTextLabel(idx, 'color', v);
            }
          }]
        })), _react["default"].createElement(_styledComponents.SidePanelSection, null, _react["default"].createElement(_styledComponents.SpaceBetweenFlexbox, null, _react["default"].createElement(_styledComponents.SBFlexboxItem, null, _react["default"].createElement(_styledComponents.PanelLabel, null, "Text anchor"), _react["default"].createElement(_itemSelector["default"], (0, _extends2["default"])({}, _layerFactory.LAYER_TEXT_CONFIGS.textAnchor, {
          selectedItems: tl.anchor,
          onChange: function onChange(val) {
            return updateLayerTextLabel(idx, 'anchor', val);
          }
        }))), _react["default"].createElement(_styledComponents.SBFlexboxItem, null, _react["default"].createElement(_styledComponents.PanelLabel, null, "Alignment"), _react["default"].createElement(_itemSelector["default"], (0, _extends2["default"])({}, _layerFactory.LAYER_TEXT_CONFIGS.textAlignment, {
          selectedItems: tl.alignment,
          onChange: function onChange(val) {
            return updateLayerTextLabel(idx, 'alignment', val);
          }
        }))))));
      }), _react["default"].createElement(_styledComponents.SidePanelSection, null, _react["default"].createElement(_styledComponents.Button, {
        link: true,
        onClick: function onClick(val) {
          return updateLayerTextLabel(textLabel.length);
        }
      }, _react["default"].createElement(_icons.Add, {
        height: "12px"
      }), "Add More Label"))));
    }
  }]);
  return TextLabelPanel;
}(_react.Component);

exports["default"] = TextLabelPanel;
(0, _defineProperty2["default"])(TextLabelPanel, "propTypes", {
  fields: _propTypes["default"].arrayOf(_propTypes["default"].object),
  textLabel: _propTypes["default"].arrayOf(_propTypes["default"].object),
  updateLayerTextLabel: _propTypes["default"].func.isRequired
});
