// kepler.gl/dist/components/side-panel/layer-panel/vis-config-by-field-selector.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = require("../../common/styled-components");

var _fieldSelector = _interopRequireDefault(require("../../common/field-selector"));

var _infoHelper = _interopRequireDefault(require("../../common/info-helper"));

var _dimensionScaleSelector = _interopRequireDefault(require("./dimension-scale-selector"));

var _utils = require("../../../utils/utils");

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
var VisConfigByFieldSelector =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(VisConfigByFieldSelector, _Component);

  function VisConfigByFieldSelector() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, VisConfigByFieldSelector);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(VisConfigByFieldSelector)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_updateVisByField", function (val) {
      _this.props.updateField(val);
    });
    return _this;
  }

  (0, _createClass2["default"])(VisConfigByFieldSelector, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          property = _this$props.property,
          showScale = _this$props.showScale,
          selectedField = _this$props.selectedField,
          description = _this$props.description,
          _this$props$scaleOpti = _this$props.scaleOptions,
          scaleOptions = _this$props$scaleOpti === void 0 ? [] : _this$props$scaleOpti;
      return _react["default"].createElement(_styledComponents.SidePanelSection, null, _react["default"].createElement(_styledComponents.SidePanelSection, null, _react["default"].createElement(_styledComponents.PanelLabelWrapper, null, _react["default"].createElement(_styledComponents.PanelLabel, null, this.props.label || "".concat((0, _utils.capitalizeFirstLetter)(property), " based on")), description && _react["default"].createElement(_infoHelper["default"], {
        description: description,
        id: "".concat(this.props.id, "-").concat(property)
      })), _react["default"].createElement(_fieldSelector["default"], {
        fields: this.props.fields,
        value: selectedField && selectedField.name,
        placeholder: this.props.placeholder,
        onSelect: this._updateVisByField,
        erasable: true
      })), _react["default"].createElement("div", null, showScale ? _react["default"].createElement(_dimensionScaleSelector["default"], {
        scaleType: this.props.scaleType,
        options: scaleOptions,
        label: "".concat(property, " scale"),
        onSelect: this.props.updateScale,
        disabled: scaleOptions.length < 2
      }) : null));
    }
  }]);
  return VisConfigByFieldSelector;
}(_react.Component);

exports["default"] = VisConfigByFieldSelector;
(0, _defineProperty2["default"])(VisConfigByFieldSelector, "propTypes", {
  channel: _propTypes["default"].string.isRequired,
  fields: _propTypes["default"].arrayOf(_propTypes["default"].any).isRequired,
  id: _propTypes["default"].string.isRequired,
  property: _propTypes["default"].string.isRequired,
  showScale: _propTypes["default"].bool.isRequired,
  updateField: _propTypes["default"].func.isRequired,
  updateScale: _propTypes["default"].func.isRequired,
  // optional
  scaleType: _propTypes["default"].string,
  selectedField: _propTypes["default"].object,
  description: _propTypes["default"].string,
  label: _propTypes["default"].string,
  placeholder: _propTypes["default"].string
});
;
