// kepler.gl/dist/components/side-panel/interaction-panel/tooltip-config.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = require("../../common/styled-components");

var _fieldSelector = _interopRequireDefault(require("../../common/field-selector"));

var _datasetTag = _interopRequireDefault(require("../common/dataset-tag"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

TooltipConfigFactory.deps = [_datasetTag["default"]];

function TooltipConfigFactory(DatasetTag) {
  var TooltipConfig = function TooltipConfig(_ref) {
    var config = _ref.config,
        datasets = _ref.datasets,
        onChange = _ref.onChange;
    return _react["default"].createElement("div", null, Object.keys(config.fieldsToShow).map(function (dataId) {
      return _react["default"].createElement(_styledComponents.SidePanelSection, {
        key: dataId
      }, _react["default"].createElement(DatasetTag, {
        dataset: datasets[dataId]
      }), _react["default"].createElement(_fieldSelector["default"], {
        fields: datasets[dataId].fields,
        value: config.fieldsToShow[dataId],
        onSelect: function onSelect(fieldsToShow) {
          var newConfig = _objectSpread({}, config, {
            fieldsToShow: _objectSpread({}, config.fieldsToShow, (0, _defineProperty2["default"])({}, dataId, fieldsToShow.map(function (d) {
              return d.name;
            })))
          });

          onChange(newConfig);
        },
        closeOnSelect: false,
        multiSelect: true,
        inputTheme: "secondary"
      }));
    }));
  };

  return TooltipConfig;
}

var _default = TooltipConfigFactory;
exports["default"] = _default;
