// kepler.gl/dist/components/common/info-helper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = require("./styled-components");

var _icons = require("./icons");

var _styledComponents2 = _interopRequireDefault(require("styled-components"));

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  align-items: center;\n  margin-left: 10px;\n  color: ", ";\n  display: inline-flex;\n\n  .info-helper__content {\n    max-width: 100px;\n  }\n\n  :hover {\n    cursor: pointer;\n    color: ", ";\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var StyledInfoHelper = _styledComponents2["default"].div(_templateObject(), function (props) {
  return props.theme.labelColor;
}, function (props) {
  return props.theme.textColorHl;
});

var propTypes = {
  description: _propTypes["default"].string.isRequired,
  containerClass: _propTypes["default"].string
};

var InfoHelper = function InfoHelper(_ref) {
  var description = _ref.description,
      containerClass = _ref.containerClass,
      id = _ref.id;
  return _react["default"].createElement(StyledInfoHelper, {
    className: "info-helper ".concat(containerClass || ''),
    "data-tip": true,
    "data-for": id
  }, _react["default"].createElement(_icons.Docs, {
    height: "16px"
  }), _react["default"].createElement(_styledComponents.Tooltip, {
    id: id,
    effect: "solid"
  }, _react["default"].createElement("div", {
    className: "info-helper__content"
  }, description)));
};

InfoHelper.propTypes = propTypes;
var _default = InfoHelper;
exports["default"] = _default;
