// kepler.gl/dist/components/modals/export-map-modal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _reactJsonPretty = _interopRequireDefault(require("react-json-pretty"));

var _userGuides = require("../../constants/user-guides");

var _icons = require("../common/icons");

var _styledComponents2 = require("../common/styled-components");

var _defaultSettings = require("../../constants/default-settings");

function _templateObject6() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  .note {\n    color: ", ";\n    font-size: 11px;\n  }\n\n  .viewer {\n    border: 1px solid ", ";\n    background-color: white;\n    border-radius: 2px;\n    display: inline-block;\n    font: inherit;\n    line-height: 1.5em;\n    padding: 0.5em 3.5em 0.5em 1em;\n    margin: 0;\n    box-sizing: border-box;\n    height: 180px;\n    width: 100%;\n    overflow-y: scroll;\n    overflow-x: auto;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    max-width: 600px;\n  }\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  height: unset;\n  width: unset;\n  img {\n    width: 180px;\n    height: 120px;\n  }\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  .disclaimer {\n    font-size: ", ";\n    color: ", ";\n    margin-top: 12px;\n  }\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  text-decoration-line: underline !important;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n  font-weight: ", ";\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  width: 100%;\n  padding: ", ";\n  color: ", ";\n  height: ", ";\n  outline: 0;\n  font-size: ", ";\n\n  :active,\n  :focus,\n  &.focus,\n  &.active {\n    outline: 0;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var NO_OP = function NO_OP() {};

var StyledInput = _styledComponents["default"].input(_templateObject(), function (props) {
  return props.theme.inputPadding;
}, function (props) {
  return props.error ? 'red' : props.theme.titleColorLT;
}, function (props) {
  return props.theme.inputBoxHeight;
}, function (props) {
  return props.theme.inputFontSize;
});

var StyledWarning = _styledComponents["default"].span(_templateObject2(), function (props) {
  return props.theme.errorColor;
}, function (props) {
  return props.theme.selectFontWeightBold;
});

var StyledLink = _styledComponents["default"].a(_templateObject3());

var INTRA_SECTION_MARGING = '8px';
var ExportMapStyledExportSection = (0, _styledComponents["default"])(_styledComponents2.StyledExportSection)(_templateObject4(), function (props) {
  return props.theme.inputFontSize;
}, function (props) {
  return props.theme.inputColor;
});

var Link = function Link(_ref) {
  var children = _ref.children,
      props = (0, _objectWithoutProperties2["default"])(_ref, ["children"]);
  return _react["default"].createElement(StyledLink, (0, _extends2["default"])({
    target: "_blank",
    rel: "noopener noreferrer"
  }, props), children);
};

var BigStyledType = (0, _styledComponents["default"])(_styledComponents2.StyledType)(_templateObject5());
var exportHtmlPropTypes = {
  options: _propTypes["default"].object,
  onEditUserMapboxAccessToken: _propTypes["default"].func.isRequired
};

var ExportHtmlMap = function ExportHtmlMap(_ref2) {
  var _ref2$onChangeExportM = _ref2.onChangeExportMapHTMLMode,
      onChangeExportMapHTMLMode = _ref2$onChangeExportM === void 0 ? NO_OP : _ref2$onChangeExportM,
      _ref2$onEditUserMapbo = _ref2.onEditUserMapboxAccessToken,
      onEditUserMapboxAccessToken = _ref2$onEditUserMapbo === void 0 ? NO_OP : _ref2$onEditUserMapbo,
      _ref2$options = _ref2.options,
      options = _ref2$options === void 0 ? {} : _ref2$options;
  return _react["default"].createElement("div", null, _react["default"].createElement(_styledComponents2.StyledExportSection, {
    style: {
      marginTop: INTRA_SECTION_MARGING
    }
  }, _react["default"].createElement("div", {
    className: "description"
  }), _react["default"].createElement("div", {
    className: "selection"
  }, "Export your map into an interactive html file.")), _react["default"].createElement(ExportMapStyledExportSection, {
    className: "export-map-modal__html-options"
  }, _react["default"].createElement("div", {
    className: "description"
  }, _react["default"].createElement("div", {
    className: "title"
  }, "Mapbox access token"), _react["default"].createElement("div", {
    className: "subtitle"
  }, "Use your own Mapbox access token in the html (optional)")), _react["default"].createElement("div", {
    className: "selection"
  }, _react["default"].createElement(StyledInput, {
    onChange: function onChange(e) {
      return onEditUserMapboxAccessToken(e.target.value);
    },
    type: "text",
    placeholder: "Paste your Mapbox access token",
    value: options ? options.userMapboxToken : ''
  }), _react["default"].createElement("div", {
    className: "disclaimer"
  }, _react["default"].createElement(StyledWarning, null, _defaultSettings.TOKEN_MISUSE_WARNING), _react["default"].createElement("span", null, _defaultSettings.DISCLAIMER), _react["default"].createElement(Link, {
    href: _userGuides.GITHUB_EXPORT_HTML_MAP
  }, "How to update an existing map token.")))), _react["default"].createElement(ExportMapStyledExportSection, null, _react["default"].createElement("div", {
    className: "description"
  }, _react["default"].createElement("div", {
    className: "title"
  }, "Map Mode"), _react["default"].createElement("div", {
    className: "subtitle"
  }, "Select the app mode. More ", _react["default"].createElement("a", {
    href: _userGuides.GITHUB_EXPORT_HTML_MAP_MODES
  }, "info"))), _react["default"].createElement("div", {
    className: "selection"
  }, _defaultSettings.EXPORT_HTML_MAP_MODE_OPTIONS.map(function (mode) {
    return _react["default"].createElement(BigStyledType, {
      key: mode.id,
      selected: options.mode === mode.id,
      available: mode.available,
      onClick: function onClick() {
        return mode.available && onChangeExportMapHTMLMode(mode.id);
      }
    }, _react["default"].createElement("img", {
      src: mode.url,
      alt: ""
    }), _react["default"].createElement("p", null, "Allow users to ", mode.label, " the map"));
  }))));
};

ExportHtmlMap.propTypes = exportHtmlPropTypes;
var StyledJsonExportSection = (0, _styledComponents["default"])(ExportMapStyledExportSection)(_templateObject6(), function (props) {
  return props.theme.errorColor;
}, function (props) {
  return props.theme.selectBorderColorLT;
});
var exportJsonPropTypes = {
  options: _propTypes["default"].object
};

var ExportJsonMap = function ExportJsonMap(_ref3) {
  var _ref3$config = _ref3.config,
      config = _ref3$config === void 0 ? {} : _ref3$config;
  return _react["default"].createElement("div", null, _react["default"].createElement(_styledComponents2.StyledExportSection, {
    style: {
      marginTop: INTRA_SECTION_MARGING
    }
  }, _react["default"].createElement("div", {
    className: "description"
  }), _react["default"].createElement("div", {
    className: "selection"
  }, "Export current map data and config into a single Json file. You can later open the same map by uploading this file to kepler.gl.")), _react["default"].createElement(StyledJsonExportSection, {
    className: "export-map-modal__json-options"
  }, _react["default"].createElement("div", {
    className: "description"
  }, _react["default"].createElement("div", {
    className: "title"
  }, "Map Config"), _react["default"].createElement("div", {
    className: "subtitle"
  }, _defaultSettings.MAP_CONFIG_DESCRIPTION, _react["default"].createElement(Link, {
    href: _userGuides.GITHUB_ADD_DATA_TO_MAP
  }, "addDataToMap"), ".")), _react["default"].createElement("div", {
    className: "selection"
  }, _react["default"].createElement("div", {
    className: "viewer"
  }, _react["default"].createElement(_reactJsonPretty["default"], {
    id: "json-pretty",
    json: config
  })), _react["default"].createElement("div", {
    className: "disclaimer"
  }, _react["default"].createElement(StyledWarning, null, "* Map config is coupled with loaded datasets. \u2018dataId\u2019 is used to bind layers, filters, and tooltips to a specific dataset. When passing this config to addDataToMap, make sure the dataset id matches the dataId/s in this config.")))));
};

ExportJsonMap.propTypes = exportJsonPropTypes;
var propTypes = {
  options: _propTypes["default"].object,
  onEditUserMapboxAccessToken: _propTypes["default"].func.isRequired,
  onChangeExportData: _propTypes["default"].func,
  onChangeExportMapType: _propTypes["default"].func,
  mapFormat: _propTypes["default"].string
};
var style = {
  width: '100%'
};

var ExportMapModal = function ExportMapModal(_ref4) {
  var _EXPORT_MAP_FORMATS$H;

  var _ref4$config = _ref4.config,
      config = _ref4$config === void 0 ? {} : _ref4$config,
      _ref4$onChangeExportD = _ref4.onChangeExportData,
      onChangeExportData = _ref4$onChangeExportD === void 0 ? NO_OP : _ref4$onChangeExportD,
      _ref4$onChangeExportM = _ref4.onChangeExportMapFormat,
      onChangeExportMapFormat = _ref4$onChangeExportM === void 0 ? NO_OP : _ref4$onChangeExportM,
      _ref4$onChangeExportM2 = _ref4.onChangeExportMapHTMLMode,
      onChangeExportMapHTMLMode = _ref4$onChangeExportM2 === void 0 ? NO_OP : _ref4$onChangeExportM2,
      _ref4$onEditUserMapbo = _ref4.onEditUserMapboxAccessToken,
      onEditUserMapboxAccessToken = _ref4$onEditUserMapbo === void 0 ? NO_OP : _ref4$onEditUserMapbo,
      _ref4$options = _ref4.options,
      options = _ref4$options === void 0 ? {} : _ref4$options;
  return _react["default"].createElement(_styledComponents2.StyledModalContent, {
    className: "export-map-modal"
  }, _react["default"].createElement("div", {
    style: style
  }, _react["default"].createElement(_styledComponents2.StyledExportSection, {
    style: {
      marginBottom: INTRA_SECTION_MARGING
    }
  }, _react["default"].createElement("div", {
    className: "description"
  }, _react["default"].createElement("div", {
    className: "title"
  }, "Map format"), _react["default"].createElement("div", {
    className: "subtitle"
  }, "Choose the format to export your map to")), _react["default"].createElement("div", {
    className: "selection"
  }, _defaultSettings.EXPORT_MAP_FORMAT_OPTIONS.map(function (op) {
    return _react["default"].createElement(_styledComponents2.StyledType, {
      key: op.id,
      selected: options.format === op.id,
      available: op.available,
      onClick: function onClick() {
        return op.available && onChangeExportMapFormat(op.id);
      }
    }, _react["default"].createElement(_icons.FileType, {
      ext: op.label,
      height: "80px",
      fontSize: "11px"
    }));
  }))), (_EXPORT_MAP_FORMATS$H = {}, (0, _defineProperty2["default"])(_EXPORT_MAP_FORMATS$H, _defaultSettings.EXPORT_MAP_FORMATS.HTML, _react["default"].createElement(ExportHtmlMap, {
    onChangeExportMapHTMLMode: onChangeExportMapHTMLMode,
    onEditUserMapboxAccessToken: onEditUserMapboxAccessToken,
    options: options[options.format]
  })), (0, _defineProperty2["default"])(_EXPORT_MAP_FORMATS$H, _defaultSettings.EXPORT_MAP_FORMATS.JSON, _react["default"].createElement(ExportJsonMap, {
    config: config,
    onChangeExportData: onChangeExportData,
    options: options[options.format]
  })), _EXPORT_MAP_FORMATS$H)[options.format]));
};

ExportMapModal.propTypes = propTypes;

var ExportMapModalFactory = function ExportMapModalFactory() {
  return ExportMapModal;
};

var _default = ExportMapModalFactory;
exports["default"] = _default;
