// kepler.gl/dist/components/modals/export-image-modal.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _exportImageUtils = require("../../utils/export-image-utils");

var _defaultSettings = require("../../constants/default-settings");

var _styledComponents2 = require("../common/styled-components");

var _switch = _interopRequireDefault(require("../common/switch"));

var _loadingSpinner = _interopRequireDefault(require("../common/loading-spinner"));

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  align-items: center;\n  display: flex;\n  flex-direction: column;\n  flex: 1;\n  justify-content: center;\n  padding: 30px;\n\n  .dimension,\n  .instruction {\n    padding: 8px 0px;\n  }\n\n  .preview-image {\n    background: #e2e2e2;\n    border-radius: 4px;\n    box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.18);\n    width: 100%;\n    padding-bottom: ", ";\n    position: relative;\n  }\n\n  .preview-image-placeholder {\n    position: absolute;\n    top: 0;\n    left: 0;\n    width: 100%;\n    height: 100%;\n  }\n\n  .preview-image-spinner {\n    position: absolute;\n    left: calc(50% - 25px);\n    top: calc(50% - 25px);\n  }\n\n  .preview-image--error {\n    font-size: 12px;\n    padding: 12px;\n    color: ", ";\n    text-align: center;\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: column;\n  justify-content: space-around;\n  width: 250px;\n\n  .image-option-section {\n    .image-option-section-title {\n      font-weight: 500;\n      font-size: 14px;\n    }\n  }\n\n  .button-list {\n    display: flex;\n    flex-direction: row;\n    padding: 8px 0px;\n  }\n\n  input {\n    margin-right: 8px;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var ImageOptionList = _styledComponents["default"].div(_templateObject());

var PreviewImageSection = _styledComponents["default"].div(_templateObject2(), function (props) {
  return props.ratio === _defaultSettings.RATIOS.SCREEN ? "".concat(100 * props.height / props.width, "%") : props.ratio === _defaultSettings.RATIOS.SIXTEEN_BY_NINE ? '56.25%' : '75%';
}, function (props) {
  return props.theme.errorColor;
});

var ExportImageModal =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(ExportImageModal, _Component);

  function ExportImageModal() {
    (0, _classCallCheck2["default"])(this, ExportImageModal);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ExportImageModal).apply(this, arguments));
  }

  (0, _createClass2["default"])(ExportImageModal, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          height = _this$props.height,
          legend = _this$props.legend,
          ratio = _this$props.ratio,
          error = _this$props.error,
          resolution = _this$props.resolution,
          width = _this$props.width,
          exporting = _this$props.exporting,
          imageDataUri = _this$props.imageDataUri,
          onChangeRatio = _this$props.onChangeRatio,
          onChangeResolution = _this$props.onChangeResolution,
          onToggleLegend = _this$props.onToggleLegend;
      var exportImageSize = (0, _exportImageUtils.calculateExportImageSize)({
        width: width,
        height: height,
        ratio: ratio,
        resolution: resolution
      });
      return _react["default"].createElement(_styledComponents2.StyledModalContent, {
        className: "export-image-modal"
      }, _react["default"].createElement(ImageOptionList, null, _react["default"].createElement("div", {
        className: "image-option-section"
      }, _react["default"].createElement("div", {
        className: "image-option-section-title"
      }, "Ratio"), "Choose the ratio for various usages.", _react["default"].createElement("div", {
        className: "button-list"
      }, _defaultSettings.RATIO_OPTIONS.map(function (op) {
        return _react["default"].createElement(_styledComponents2.SelectionButton, {
          key: op.id,
          selected: ratio === op.id,
          onClick: function onClick() {
            return onChangeRatio({
              ratio: op.id
            });
          }
        }, op.label);
      }))), _react["default"].createElement("div", {
        className: "image-option-section"
      }, _react["default"].createElement("div", {
        className: "image-option-section-title"
      }, "Resolution"), "High resolution is better for prints.", _react["default"].createElement("div", {
        className: "button-list"
      }, _defaultSettings.RESOLUTION_OPTIONS.map(function (op) {
        return _react["default"].createElement(_styledComponents2.SelectionButton, {
          key: op.id,
          selected: resolution === op.id,
          onClick: function onClick() {
            return op.available && onChangeResolution({
              resolution: op.id
            });
          }
        }, op.label);
      }))), _react["default"].createElement("div", {
        className: "image-option-section"
      }, _react["default"].createElement("div", {
        className: "image-option-section-title"
      }, "Map Legend"), _react["default"].createElement(_switch["default"], {
        type: "checkbox",
        id: "add-map-legend",
        checked: legend,
        label: "Add legend on map",
        onChange: onToggleLegend
      }))), _react["default"].createElement(PreviewImageSection, {
        ratio: ratio,
        width: width,
        height: height
      }, _react["default"].createElement("div", {
        className: "dimension"
      }, "".concat(exportImageSize.width, " x ").concat(exportImageSize.height)), _react["default"].createElement("div", {
        className: "preview-image"
      }, exporting ? _react["default"].createElement("div", {
        className: "preview-image-spinner"
      }, _react["default"].createElement(_loadingSpinner["default"], null)) : error ? _react["default"].createElement("div", {
        className: "preview-image--error"
      }, _react["default"].createElement("span", null, " ", error.message || 'Generate map image failed!')) : _react["default"].createElement("img", {
        className: "preview-image-placeholder",
        src: imageDataUri
      }))));
    }
  }]);
  return ExportImageModal;
}(_react.Component);

(0, _defineProperty2["default"])(ExportImageModal, "propTypes", {
  height: _propTypes["default"].number.isRequired,
  ratio: _propTypes["default"].string.isRequired,
  resolution: _propTypes["default"].string.isRequired,
  width: _propTypes["default"].number.isRequired,
  exporting: _propTypes["default"].bool.isRequired,
  imageDataUri: _propTypes["default"].string,
  // callbacks
  onChangeRatio: _propTypes["default"].func.isRequired,
  onChangeResolution: _propTypes["default"].func.isRequired,
  onToggleLegend: _propTypes["default"].func.isRequired
});

var ExportImageModalFactory = function ExportImageModalFactory() {
  return ExportImageModal;
};

var _default = ExportImageModalFactory;
exports["default"] = _default;
