// kepler.gl/dist/components/common/animation-control/floating-time-display.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = FloatingTimeDisplayFactory;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _moment = _interopRequireDefault(require("moment"));

var _reselect = require("reselect");

var _icons = require("../icons");

var _defaultSettings = require("../../../constants/default-settings");

var _styledComponents2 = require("../styled-components");

function _templateObject7() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  margin: 0 12px;\n"]);

  _templateObject7 = function _templateObject7() {
    return data;
  };

  return data;
}

function _templateObject6() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: column;\n"]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n  display: flex;\n  font-size: 14px;\n  font-weight: 500;\n  justify-content: center;\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n  display: flex;\n  font-size: 12px;\n  font-weight: 500;\n  justify-content: center;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: column;\n  justify-content: center;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  align-items: center;\n  display: flex;\n  flex-direction: row;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  background-color: ", ";\n  border-radius: ", "px;\n  bottom: ", ";\n  color: ", ";\n  display: flex;\n  height: ", "px;\n  justify-content: center;\n  left: calc(50% - 88px);\n  min-width: ", "px;\n  opacity: ", ";\n  padding: ", ";\n  position: absolute;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var StyledTimeDisplay = _styledComponents["default"].div(_templateObject(), function (props) {
  return props.theme.panelBackground;
}, function (props) {
  return props.theme.timeDisplayBorderRadius;
}, function (props) {
  return "calc(100% + ".concat(props.theme.bottomPanelGap, "px)");
}, function (props) {
  return props.theme.titleTextColor;
}, function (props) {
  return props.theme.timeDisplayHeight;
}, function (props) {
  return props.theme.timeDisplayMinWidth;
}, function (props) {
  return props.theme.timeDisplayOpacity;
}, function (props) {
  return props.theme.timeDisplayPadding;
});

var StyledTimeDisplayGroups = _styledComponents["default"].div(_templateObject2());

var StyledTimeDisplayRows = _styledComponents["default"].div(_templateObject3());

var StyledTimeDisplayTop = _styledComponents["default"].div.attrs({
  className: 'animation-control__time-display__top'
})(_templateObject4(), function (props) {
  return props.theme.textColor;
});

var StyledTimeDisplayBottom = _styledComponents["default"].div.attrs({
  className: 'animation-control__time-display__bottom'
})(_templateObject5(), function (props) {
  return props.theme.titleTextColor;
});

var StyledTimeValueGroup = _styledComponents["default"].div.attrs({
  className: 'animation-control__time-value-group'
})(_templateObject6());

var StyledHorizontalBar = _styledComponents["default"].div.attrs({
  className: 'animation-control__horizontal-bar'
})(_templateObject7());

var TimeDivider = function TimeDivider() {
  return _react["default"].createElement(StyledHorizontalBar, null, _react["default"].createElement(_icons.Minus, {
    height: "12px"
  }));
};

var TimeDisplayRow = function TimeDisplayRow(_ref) {
  var _ref$timeValues = _ref.timeValues,
      timeValues = _ref$timeValues === void 0 ? [] : _ref$timeValues;
  return _react["default"].createElement(_styledComponents2.CenterFlexbox, null, _react["default"].createElement("div", null, timeValues[0]), timeValues[1] ? _react["default"].createElement(TimeDivider, null) : null, timeValues[1] ? _react["default"].createElement("div", null, timeValues[1]) : null);
};

function FloatingTimeDisplayFactory() {
  var FloatingTimeDisplay =
  /*#__PURE__*/
  function (_PureComponent) {
    (0, _inherits2["default"])(FloatingTimeDisplay, _PureComponent);

    function FloatingTimeDisplay() {
      var _getPrototypeOf2;

      var _this;

      (0, _classCallCheck2["default"])(this, FloatingTimeDisplay);

      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(FloatingTimeDisplay)).call.apply(_getPrototypeOf2, [this].concat(args)));
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "timeSelector", function (props) {
        return props.currentTime;
      });
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "formatSelector", function (props) {
        return props.format;
      });
      (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "displayTimeSelector", (0, _reselect.createSelector)(_this.timeSelector, _this.formatSelector, function (currentTime, format) {
        var groupTime = Array.isArray(currentTime) ? currentTime : [currentTime];
        return groupTime.reduce(function (accu, curr) {
          var displayDateTime = _moment["default"].utc(curr).format(format);

          var _displayDateTime$spli = displayDateTime.split(' '),
              _displayDateTime$spli2 = (0, _slicedToArray2["default"])(_displayDateTime$spli, 2),
              displayDate = _displayDateTime$spli2[0],
              displayTime = _displayDateTime$spli2[1];

          if (!accu.displayDate.includes(displayDate)) {
            accu.displayDate.push(displayDate);
          }

          accu.displayTime.push(displayTime);
          return accu;
        }, {
          displayDate: [],
          displayTime: []
        });
      }));
      return _this;
    }

    (0, _createClass2["default"])(FloatingTimeDisplay, [{
      key: "render",
      value: function render() {
        var _this$displayTimeSele = this.displayTimeSelector(this.props),
            displayDate = _this$displayTimeSele.displayDate,
            displayTime = _this$displayTimeSele.displayTime;

        var twoGroups = displayDate.length === 2 && displayTime.length === 2;
        var bottomRow = displayTime.length ? displayTime : displayDate.length ? displayDate : null;
        var topRow = displayDate.length && displayTime.length ? displayDate : null;
        return _react["default"].createElement(StyledTimeDisplay, {
          className: "animation-control__time-display"
        }, twoGroups ? _react["default"].createElement(StyledTimeDisplayGroups, null, _react["default"].createElement(StyledTimeValueGroup, null, _react["default"].createElement(StyledTimeDisplayTop, null, displayDate[0]), _react["default"].createElement(StyledTimeDisplayBottom, null, displayTime[0])), _react["default"].createElement(TimeDivider, null), _react["default"].createElement(StyledTimeValueGroup, null, _react["default"].createElement(StyledTimeDisplayTop, null, displayDate[1]), _react["default"].createElement(StyledTimeDisplayBottom, null, displayTime[1]))) : _react["default"].createElement(StyledTimeDisplayRows, null, topRow ? _react["default"].createElement(StyledTimeDisplayTop, null, _react["default"].createElement(TimeDisplayRow, {
          timeValues: topRow
        })) : null, bottomRow ? _react["default"].createElement(StyledTimeDisplayBottom, null, _react["default"].createElement(TimeDisplayRow, {
          timeValues: bottomRow
        })) : null));
      }
    }]);
    return FloatingTimeDisplay;
  }(_react.PureComponent);

  FloatingTimeDisplay.defaultProps = {
    format: _defaultSettings.DEFAULT_TIME_FORMAT,
    currentTime: null
  };
  return FloatingTimeDisplay;
}
