// kepler.gl/dist/components/common/animation-control/playback-controls.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _classnames = _interopRequireDefault(require("classnames"));

var _styledComponents2 = require("../styled-components");

var _icons = require("../icons");

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  padding: 6px 4px;\n  svg {\n    margin: 0 6px;\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  margin-right: 12px;\n\n  &.disabled {\n    opacity: 0.4;\n    pointer-events: none;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var StyledAnimationControls = _styledComponents["default"].div(_templateObject());

var IconButton = (0, _styledComponents["default"])(_styledComponents2.Button)(_templateObject2());

function nop() {}

;
var DEFAULT_BUTTON_HEIGHT = '18px';

function AnimationPlaybacksFactory() {
  var AnimationPlaybacks = function AnimationPlaybacks(_ref) {
    var isAnimatable = _ref.isAnimatable,
        isAnimating = _ref.isAnimating,
        buttonStyle = _ref.buttonStyle,
        _ref$pauseAnimation = _ref.pauseAnimation,
        pauseAnimation = _ref$pauseAnimation === void 0 ? nop : _ref$pauseAnimation,
        _ref$updateAnimationT = _ref.updateAnimationTime,
        updateAnimationTime = _ref$updateAnimationT === void 0 ? nop : _ref$updateAnimationT,
        _ref$startAnimation = _ref.startAnimation,
        startAnimation = _ref$startAnimation === void 0 ? nop : _ref$startAnimation,
        _ref$buttonHeight = _ref.buttonHeight,
        buttonHeight = _ref$buttonHeight === void 0 ? DEFAULT_BUTTON_HEIGHT : _ref$buttonHeight;
    var btnStyle = buttonStyle ? (0, _defineProperty2["default"])({}, buttonStyle, true) : {};
    return _react["default"].createElement(StyledAnimationControls, {
      className: (0, _classnames["default"])('time-range-slider__control', {
        disabled: !isAnimatable
      })
    }, _react["default"].createElement(_styledComponents2.ButtonGroup, null, _react["default"].createElement(IconButton, (0, _extends2["default"])({
      className: "playback-control-button"
    }, btnStyle, {
      onClick: updateAnimationTime
    }), _react["default"].createElement(_icons.Reset, {
      height: buttonHeight
    })), _react["default"].createElement(IconButton, (0, _extends2["default"])({}, btnStyle, {
      className: (0, _classnames["default"])('playback-control-button', {
        active: isAnimating
      }),
      onClick: isAnimating ? pauseAnimation : startAnimation
    }), isAnimating ? _react["default"].createElement(_icons.Pause, {
      height: buttonHeight
    }) : _react["default"].createElement(_icons.Play, {
      height: buttonHeight
    }))));
  };

  return AnimationPlaybacks;
}

var _default = AnimationPlaybacksFactory;
exports["default"] = _default;
