// kepler.gl/dist/components/side-panel/panel-header.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.SaveExportDropdownFactory = exports.SaveMapFactory = exports.ExportMapFactory = exports.ExportDataFactory = exports.ExportImageFactory = exports.PanelAction = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));

var _react = _interopRequireWildcard(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents2 = require("../common/styled-components");

var _logo = _interopRequireDefault(require("../common/logo"));

var _icons = require("../common/icons");

var _panelDropdown = _interopRequireDefault(require("./panel-dropdown"));

function _templateObject5() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  background-color: ", ";\n  box-shadow: ", ";\n  font-size: 11px;\n  padding: 16px 0;\n  position: absolute;\n  left: 64px;\n  transition: ", ";\n  display: flex;\n  margin-top: ", ";\n  opacity: ", ";\n  transform: translateX(calc(-50% + 20px));\n  pointer-events:  ", ";\n  z-index: 1000;\n\n  .save-export-dropdown__inner {\n    box-shadow: none;\n    background-color: transparent;\n    display: flex;\n  }\n\n  .save-export-dropdown__item {\n    align-items: center;\n    border-right: 1px solid ", ";\n    color: ", ";\n    display: flex;\n    flex-direction: column;\n    padding: 0 22px;\n\n    :hover {\n      cursor: pointer;\n      color: ", ";\n    }\n\n    &:last-child {\n      border-right: 0;\n    }\n  }\n\n  .save-export-dropdown__title {\n    white-space: nowrap;\n    margin-top: 4px;\n  }\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  align-items: center;\n  border-radius: 2px;\n  color: ", ";\n  display: flex;\n  height: 26px;\n  justify-content: space-between;\n  margin-left: 4px;\n  width: 70px;\n  padding: 5px;\n  font-weight: bold;\n  a {\n    height: 20px;\n  }\n\n  :hover {\n    cursor: pointer;\n    background-color: ", ";\n    color: ", ";\n\n    a {\n      color: ", ";\n    }\n  }\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  justify-content: space-between;\n  margin-bottom: 16px;\n  width: 100%;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = (0, _taggedTemplateLiteral2["default"])(["\n  background-color: ", ";\n  padding: 12px 16px 0 16px;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

var StyledPanelHeader = _styledComponents["default"].div.attrs({
  className: 'side-side-panel__header'
})(_templateObject(), function (props) {
  return props.theme.sidePanelHeaderBg;
});

var StyledPanelHeaderTop = _styledComponents["default"].div.attrs({
  className: 'side-panel__header__top'
})(_templateObject2());

var StyledPanelTopActions = _styledComponents["default"].div.attrs({
  className: 'side-panel__header__actions'
})(_templateObject3());

var StyledPanelAction = _styledComponents["default"].div.attrs({
  className: 'side-panel__header__actions'
})(_templateObject4(), function (props) {
  return props.active ? props.theme.textColorHl : props.theme.subtextColor;
}, function (props) {
  return props.theme.secondaryBtnActBgd;
}, function (props) {
  return props.theme.textColorHl;
}, function (props) {
  return props.theme.textColorHl;
});

var StyledPanelDropdown = _styledComponents["default"].div(_templateObject5(), function (props) {
  return props.theme.dropdownListBgd;
}, function (props) {
  return props.theme.dropdownListShadow;
}, function (props) {
  return props.theme.transitionSlow;
}, function (props) {
  return props.show ? '6px' : '20px';
}, function (props) {
  return props.show ? 1 : 0;
}, function (props) {
  return props.show ? 'all' : 'none';
}, function (props) {
  return props.theme.panelHeaderIcon;
}, function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.theme.textColorHl;
});

var PanelAction = function PanelAction(_ref) {
  var item = _ref.item,
      onClick = _ref.onClick;
  return _react["default"].createElement(StyledPanelAction, {
    className: "side-panel__panel-header__action",
    "data-tip": true,
    "data-for": "".concat(item.id, "-action"),
    onClick: onClick
  }, item.label ? _react["default"].createElement("p", null, item.label) : null, _react["default"].createElement("a", {
    target: item.blank ? '_blank' : '',
    href: item.href
  }, _react["default"].createElement(item.iconComponent, {
    height: "20px"
  })), item.tooltip ? _react["default"].createElement(_styledComponents2.Tooltip, {
    id: "".concat(item.id, "-action"),
    place: "bottom",
    delayShow: 500,
    effect: "solid"
  }, _react["default"].createElement("span", null, item.tooltip)) : null);
};

exports.PanelAction = PanelAction;

var PanelItem = function PanelItem(_ref2) {
  var onClose = _ref2.onClose,
      onClickHandler = _ref2.onClickHandler,
      label = _ref2.label,
      icon = _ref2.icon;
  return _react["default"].createElement("div", {
    className: "save-export-dropdown__item",
    onClick: function onClick(e) {
      e.stopPropagation();
      onClose();
      onClickHandler();
    }
  }, icon, _react["default"].createElement("div", {
    className: "save-export-dropdown__title"
  }, label));
};

var ExportImageFactory = function ExportImageFactory() {
  var ExportImage = function ExportImage(props) {
    return _react["default"].createElement(PanelItem, props);
  };

  ExportImage.defaultProps = {
    label: 'Export Image',
    icon: _react["default"].createElement(_icons.Picture, null)
  };
  return ExportImage;
};

exports.ExportImageFactory = ExportImageFactory;

var ExportDataFactory = function ExportDataFactory() {
  var ExportData = function ExportData(props) {
    return _react["default"].createElement(PanelItem, props);
  };

  ExportData.defaultProps = {
    label: 'Export Data',
    icon: _react["default"].createElement(_icons.Files, null)
  };
  return ExportData;
};

exports.ExportDataFactory = ExportDataFactory;

var ExportMapFactory = function ExportMapFactory() {
  var ExportMap = function ExportMap(props) {
    return _react["default"].createElement(PanelItem, props);
  };

  ExportMap.defaultProps = {
    label: 'Export Map',
    icon: _react["default"].createElement(_icons.Map, null)
  };
  return ExportMap;
};

exports.ExportMapFactory = ExportMapFactory;

var SaveMapFactory = function SaveMapFactory() {
  var SaveMap = function SaveMap(props) {
    return _react["default"].createElement(PanelItem, props);
  };

  SaveMap.defaultProps = {
    label: 'Save Map',
    icon: _react["default"].createElement(_icons.Share, null)
  };
  return SaveMap;
};

exports.SaveMapFactory = SaveMapFactory;

var SaveExportDropdownFactory = function SaveExportDropdownFactory(ExportImage, ExportData, ExportMap, SaveMap) {
  var SaveExportDropdown = function SaveExportDropdown(_ref3) {
    var onExportImage = _ref3.onExportImage,
        onExportData = _ref3.onExportData,
        onExportConfig = _ref3.onExportConfig,
        onExportMap = _ref3.onExportMap,
        onSaveMap = _ref3.onSaveMap,
        show = _ref3.show,
        onClose = _ref3.onClose;
    return _react["default"].createElement(StyledPanelDropdown, {
      show: show,
      className: "save-export-dropdown"
    }, _react["default"].createElement(_panelDropdown["default"], {
      className: "save-export-dropdown__inner",
      show: show,
      onClose: onClose
    }, _react["default"].createElement(ExportImage, {
      onClickHandler: onExportImage,
      onClose: onClose
    }), _react["default"].createElement(ExportData, {
      onClickHandler: onExportData,
      onClose: onClose
    }), _react["default"].createElement(ExportMap, {
      onClickHandler: onExportMap,
      onClose: onClose
    }), onSaveMap ? _react["default"].createElement(SaveMap, {
      onClickHandler: onSaveMap,
      onClose: onClose
    }) : null));
  };

  return SaveExportDropdown;
};

exports.SaveExportDropdownFactory = SaveExportDropdownFactory;
SaveExportDropdownFactory.deps = [ExportImageFactory, ExportDataFactory, ExportMapFactory, SaveMapFactory];
PanelHeaderFactory.deps = [SaveExportDropdownFactory];

function PanelHeaderFactory(SaveExportDropdown) {
  var _class, _temp;

  return _temp = _class =
  /*#__PURE__*/
  function (_Component) {
    (0, _inherits2["default"])(PanelHeader, _Component);

    function PanelHeader() {
      (0, _classCallCheck2["default"])(this, PanelHeader);
      return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(PanelHeader).apply(this, arguments));
    }

    (0, _createClass2["default"])(PanelHeader, [{
      key: "render",
      value: function render() {
        var _this$props = this.props,
            appName = _this$props.appName,
            version = _this$props.version,
            actionItems = _this$props.actionItems,
            onSaveMap = _this$props.onSaveMap,
            onExportImage = _this$props.onExportImage,
            onExportData = _this$props.onExportData,
            onExportConfig = _this$props.onExportConfig,
            onExportMap = _this$props.onExportMap,
            visibleDropdown = _this$props.visibleDropdown,
            showExportDropdown = _this$props.showExportDropdown,
            hideExportDropdown = _this$props.hideExportDropdown;
        return _react["default"].createElement(StyledPanelHeader, {
          className: "side-panel__panel-header"
        }, _react["default"].createElement(StyledPanelHeaderTop, {
          className: "side-panel__panel-header__top"
        }, _react["default"].createElement(this.props.logoComponent, {
          appName: appName,
          version: version
        }), _react["default"].createElement(StyledPanelTopActions, null, actionItems.map(function (item) {
          return _react["default"].createElement("div", {
            className: "side-panel__panel-header__right",
            key: item.id,
            style: {
              position: 'relative'
            }
          }, _react["default"].createElement(PanelAction, {
            item: item,
            onClick: function onClick() {
              if (item.dropdownComponent) {
                showExportDropdown(item.id);
              }

              item.onClick();
            }
          }), item.dropdownComponent ? _react["default"].createElement(item.dropdownComponent, {
            onClose: hideExportDropdown,
            show: visibleDropdown === item.id,
            onSaveMap: onSaveMap,
            onExportData: onExportData,
            onExportImage: onExportImage,
            onExportConfig: onExportConfig,
            onExportMap: onExportMap
          }) : null);
        }))));
      }
    }]);
    return PanelHeader;
  }(_react.Component), (0, _defineProperty2["default"])(_class, "propTypes", {
    appName: _propTypes["default"].string,
    version: _propTypes["default"].string,
    uiState: _propTypes["default"].object,
    uiStateActions: _propTypes["default"].object,
    logoComponent: _propTypes["default"].oneOfType([_propTypes["default"].element, _propTypes["default"].func]),
    actionItems: _propTypes["default"].arrayOf(_propTypes["default"].any)
  }), (0, _defineProperty2["default"])(_class, "defaultProps", {
    logoComponent: _logo["default"],
    actionItems: [{
      id: 'save',
      iconComponent: _icons.Save,
      onClick: function onClick() {},
      label: 'Share',
      dropdownComponent: SaveExportDropdown
    }]
  }), _temp;
}

var _default = PanelHeaderFactory;
exports["default"] = _default;
