// kepler.gl/dist/deckgl-layers/cluster-layer/cluster-layer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _deck = require("deck.gl");

var _geoViewport = _interopRequireDefault(require("@mapbox/geo-viewport"));

var _d3Array = require("d3-array");

var _dataScaleUtils = require("../../utils/data-scale-utils");

var _utils = require("../layer-utils/utils");

var _colorRanges = require("../../constants/color-ranges");

var _layerFactory = require("../../layers/layer-factory");

var _defaultSettings = require("../../constants/default-settings");

var _clusterUtils = require("../layer-utils/cluster-utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var defaultRadius = _layerFactory.LAYER_VIS_CONFIGS.clusterRadius.defaultValue;
var defaultRadiusRange = _layerFactory.LAYER_VIS_CONFIGS.clusterRadiusRange.defaultValue;
var defaultProps = {
  clusterRadius: defaultRadius,
  colorDomain: null,
  colorRange: _colorRanges.DefaultColorRange,
  colorScale: _defaultSettings.SCALE_TYPES.quantize,
  radiusRange: defaultRadiusRange,
  // maybe later...
  lowerPercentile: 0,
  upperPercentile: 100,
  getPosition: function getPosition(x) {
    return x.position;
  },
  // if want to have color based on customized aggregator, instead of count
  getColorValue: function getColorValue(points) {
    return points.length;
  },
  //  if want to have radius based on customized aggregator, instead of count
  getRadiusValue: function getRadiusValue(cell) {
    return cell.properties.point_count;
  },
  fp64: false
};

var ClusterLayer =
/*#__PURE__*/
function (_CompositeLayer) {
  (0, _inherits2["default"])(ClusterLayer, _CompositeLayer);

  function ClusterLayer() {
    (0, _classCallCheck2["default"])(this, ClusterLayer);
    return (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(ClusterLayer).apply(this, arguments));
  }

  (0, _createClass2["default"])(ClusterLayer, [{
    key: "initializeState",
    value: function initializeState() {
      this.state = {
        clusters: null,
        geoJSON: null
      };
    }
  }, {
    key: "shouldUpdateState",
    value: function shouldUpdateState(_ref) {
      var changeFlags = _ref.changeFlags;
      return changeFlags.somethingChanged;
    }
  }, {
    key: "updateState",
    value: function updateState(_ref2) {
      var context = _ref2.context,
          oldProps = _ref2.oldProps,
          props = _ref2.props,
          changeFlags = _ref2.changeFlags;

      if (changeFlags.dataChanged || this.needsReProjectPoints(oldProps, props)) {
        // project data into clusters, and get clustered data
        this.processGeoJSON();
        this.getClusters(); // this needs clustered data to be set

        this.getColorValueDomain();
      } else if (this.needsReclusterPoints(oldProps, props)) {
        this.getClusters();
        this.getColorValueDomain();
      } else if (this.needsRecalculateScaleFunction(oldProps, props)) {
        this.getColorValueDomain();
      }
    }
  }, {
    key: "needsReProjectPoints",
    value: function needsReProjectPoints(oldProps, props) {
      return oldProps.clusterRadius !== props.clusterRadius || oldProps.getPosition !== props.getPosition;
    }
  }, {
    key: "needsReclusterPoints",
    value: function needsReclusterPoints(oldProps, props) {
      return Math.round(oldProps.zoom) !== Math.round(props.zoom);
    }
  }, {
    key: "needsRecalculateScaleFunction",
    value: function needsRecalculateScaleFunction(oldProps, props) {
      return (0, _utils.needsRecalculateColorDomain)(oldProps, props) || (0, _utils.needReCalculateScaleFunction)(oldProps, props) || (0, _utils.needsRecalculateRadiusRange)(oldProps, props) || oldProps.getColorValue !== props.getColorValue;
    }
  }, {
    key: "processGeoJSON",
    value: function processGeoJSON() {
      var _this$props = this.props,
          data = _this$props.data,
          getPosition = _this$props.getPosition;
      this.setState({
        geoJSON: (0, _clusterUtils.getGeoJSON)(data, getPosition)
      });
      (0, _clusterUtils.clearClustererCache)();
    }
  }, {
    key: "getClusters",
    value: function getClusters() {
      var geoJSON = this.state.geoJSON;
      var clusterRadius = this.props.clusterRadius;
      var _this$context = this.context,
          viewport = _this$context.viewport,
          _this$context$viewpor = _this$context.viewport,
          longitude = _this$context$viewpor.longitude,
          latitude = _this$context$viewpor.latitude,
          height = _this$context$viewpor.height,
          width = _this$context$viewpor.width; // zoom needs to be an integer for the different map utils. Also helps with cache key.

      var zoom = Math.round(viewport.zoom);

      var bbox = _geoViewport["default"].bounds([longitude, latitude], zoom, [width, height]);

      var clusters = (0, _clusterUtils.clustersAtZoom)({
        bbox: bbox,
        clusterRadius: clusterRadius,
        geoJSON: geoJSON,
        zoom: zoom
      });
      this.setState({
        clusters: clusters
      });
    }
  }, {
    key: "getColorValueDomain",
    value: function getColorValueDomain() {
      var _this$props2 = this.props,
          colorScale = _this$props2.colorScale,
          getColorValue = _this$props2.getColorValue,
          getRadiusValue = _this$props2.getRadiusValue,
          onSetColorDomain = _this$props2.onSetColorDomain;
      var clusters = this.state.clusters;
      var radiusDomain = [0, (0, _d3Array.max)(clusters, getRadiusValue)];
      var colorValues = clusters.map(function (d) {
        return getColorValue(d.properties.points);
      });

      var identity = function identity(d) {
        return d;
      };

      var colorDomain = colorScale === _defaultSettings.SCALE_TYPES.ordinal ? (0, _dataScaleUtils.getOrdinalDomain)(colorValues, identity) : colorScale === _defaultSettings.SCALE_TYPES.quantile ? (0, _dataScaleUtils.getQuantileDomain)(colorValues, identity, _d3Array.ascending) : (0, _dataScaleUtils.getLinearDomain)(colorValues, identity);
      this.setState({
        colorDomain: colorDomain,
        radiusDomain: radiusDomain
      });
      (0, _utils.getColorScaleFunction)(this);
      (0, _utils.getRadiusScaleFunction)(this);
      onSetColorDomain(colorDomain);
    }
  }, {
    key: "getUpdateTriggers",
    value: function getUpdateTriggers() {
      return {
        getColor: {
          colorRange: this.props.colorRange,
          colorDomain: this.props.colorDomain,
          getColorValue: this.props.getColorValue,
          colorScale: this.props.colorScale,
          lowerPercentile: this.props.lowerPercentile,
          upperPercentile: this.props.upperPercentile
        },
        getRadius: {
          radiusRange: this.props.radiusRange,
          radiusDomain: this.props.radiusDomain,
          getRadiusValue: this.props.getRadiusValue
        }
      };
    }
    /*
     * override default layer method to calculate cell color based on color scale function
     */

  }, {
    key: "_onGetSublayerColor",
    value: function _onGetSublayerColor(cell) {
      var getColorValue = this.props.getColorValue;
      var _this$state = this.state,
          colorScaleFunc = _this$state.colorScaleFunc,
          colorDomain = _this$state.colorDomain;
      var cv = getColorValue(cell.properties.points); // if cell value is outside domain, set alpha to 0

      var color = cv >= colorDomain[0] && cv <= colorDomain[colorDomain.length - 1] ? colorScaleFunc(cv) : [0, 0, 0, 0]; // add final alpha to color

      color[3] = Number.isFinite(color[3]) ? color[3] : 255;
      return color;
    }
  }, {
    key: "_onGetSublayerRadius",
    value: function _onGetSublayerRadius(cell) {
      var getRadiusValue = this.props.getRadiusValue;
      var radiusScaleFunc = this.state.radiusScaleFunc;
      return radiusScaleFunc(getRadiusValue(cell));
    }
  }, {
    key: "getPickingInfo",
    value: function getPickingInfo(_ref3) {
      var info = _ref3.info;
      var clusters = this.state.clusters;
      var isPicked = info.picked && info.index > -1;
      var object = null;

      if (isPicked) {
        // add cluster colorValue to object
        var cluster = clusters[info.index];
        var colorValue = this.props.getColorValue(cluster.properties.points);
        object = _objectSpread({}, cluster.properties, {
          colorValue: colorValue,
          radius: this._onGetSublayerRadius(cluster),
          position: cluster.geometry.coordinates
        });
      }

      return _objectSpread({}, info, {
        picked: Boolean(object),
        // override object with picked cluster property
        object: object
      });
    }
  }, {
    key: "renderLayers",
    value: function renderLayers() {
      // for subclassing, override this method to return
      // customized sub layer props
      var _this$props3 = this.props,
          id = _this$props3.id,
          radiusScale = _this$props3.radiusScale,
          fp64 = _this$props3.fp64; // base layer props

      var _this$props4 = this.props,
          opacity = _this$props4.opacity,
          pickable = _this$props4.pickable,
          autoHighlight = _this$props4.autoHighlight,
          highlightColor = _this$props4.highlightColor; // return props to the sublayer constructor

      return new _deck.ScatterplotLayer({
        id: "".concat(id, "-cluster"),
        data: this.state.clusters,
        radiusScale: radiusScale,
        fp64: fp64,
        opacity: opacity,
        pickable: pickable,
        autoHighlight: autoHighlight,
        highlightColor: highlightColor,
        getPosition: function getPosition(d) {
          return d.geometry.coordinates;
        },
        getRadius: this._onGetSublayerRadius.bind(this),
        getColor: this._onGetSublayerColor.bind(this),
        updateTriggers: this.getUpdateTriggers()
      });
    }
  }]);
  return ClusterLayer;
}(_deck.CompositeLayer);

exports["default"] = ClusterLayer;
ClusterLayer.layerName = 'ClusterLayer';
ClusterLayer.defaultProps = defaultProps;
