// kepler.gl/dist/constants/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "KEPLER_GL_VERSION", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.KEPLER_GL_VERSION;
  }
});
Object.defineProperty(exports, "KEPLER_GL_WEBSITE", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.KEPLER_GL_WEBSITE;
  }
});
Object.defineProperty(exports, "KEPLER_GL_NAME", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.KEPLER_GL_NAME;
  }
});
Object.defineProperty(exports, "DIMENSIONS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.DIMENSIONS;
  }
});
Object.defineProperty(exports, "ALL_FIELD_TYPES", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.ALL_FIELD_TYPES;
  }
});
Object.defineProperty(exports, "FIELD_OPTS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.FIELD_OPTS;
  }
});
Object.defineProperty(exports, "FILTER_TYPES", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.FILTER_TYPES;
  }
});
Object.defineProperty(exports, "GEOJSON_FIELDS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.GEOJSON_FIELDS;
  }
});
Object.defineProperty(exports, "ICON_FIELDS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.ICON_FIELDS;
  }
});
Object.defineProperty(exports, "TRIP_POINT_FIELDS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.TRIP_POINT_FIELDS;
  }
});
Object.defineProperty(exports, "TRIP_ARC_FIELDS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.TRIP_ARC_FIELDS;
  }
});
Object.defineProperty(exports, "SCALE_TYPES", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.SCALE_TYPES;
  }
});
Object.defineProperty(exports, "LAYER_TYPES", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.LAYER_TYPES;
  }
});
Object.defineProperty(exports, "LAYER_BLENDINGS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.LAYER_BLENDINGS;
  }
});
Object.defineProperty(exports, "AGGREGATION_TYPES", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.AGGREGATION_TYPES;
  }
});
Object.defineProperty(exports, "MAX_DEFAULT_TOOLTIPS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.MAX_DEFAULT_TOOLTIPS;
  }
});
Object.defineProperty(exports, "DATA_TABLE_ID", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.DATA_TABLE_ID;
  }
});
Object.defineProperty(exports, "DELETE_DATA_ID", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.DELETE_DATA_ID;
  }
});
Object.defineProperty(exports, "ADD_DATA_ID", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.ADD_DATA_ID;
  }
});
Object.defineProperty(exports, "EXPORT_IMAGE_ID", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.EXPORT_IMAGE_ID;
  }
});
Object.defineProperty(exports, "EXPORT_DATA_ID", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.EXPORT_DATA_ID;
  }
});
Object.defineProperty(exports, "EXPORT_MAP_ID", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.EXPORT_MAP_ID;
  }
});
Object.defineProperty(exports, "ADD_MAP_STYLE_ID", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.ADD_MAP_STYLE_ID;
  }
});
Object.defineProperty(exports, "DEFAULT_LAYER_GROUPS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.DEFAULT_LAYER_GROUPS;
  }
});
Object.defineProperty(exports, "DEFAULT_MAP_STYLES", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.DEFAULT_MAP_STYLES;
  }
});
Object.defineProperty(exports, "THEME", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.THEME;
  }
});
Object.defineProperty(exports, "EXPORT_MAP_FORMATS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.EXPORT_MAP_FORMATS;
  }
});
Object.defineProperty(exports, "EXPORT_MAP_FORMAT_OPTIONS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.EXPORT_MAP_FORMAT_OPTIONS;
  }
});
Object.defineProperty(exports, "EXPORT_DATA_TYPE_OPTIONS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.EXPORT_DATA_TYPE_OPTIONS;
  }
});
Object.defineProperty(exports, "DEFAULT_NOTIFICATION_TYPES", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.DEFAULT_NOTIFICATION_TYPES;
  }
});
Object.defineProperty(exports, "DEFAULT_NOTIFICATION_TOPICS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.DEFAULT_NOTIFICATION_TOPICS;
  }
});
Object.defineProperty(exports, "DEFAULT_COLOR_RANGE", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.DEFAULT_COLOR_RANGE;
  }
});
Object.defineProperty(exports, "PANELS", {
  enumerable: true,
  get: function get() {
    return _defaultSettings.PANELS;
  }
});
Object.defineProperty(exports, "VizColorPalette", {
  enumerable: true,
  get: function get() {
    return _customColorRanges.VizColorPalette;
  }
});
Object.defineProperty(exports, "DataVizColors", {
  enumerable: true,
  get: function get() {
    return _customColorRanges.DataVizColors;
  }
});
Object.defineProperty(exports, "COLOR_RANGES", {
  enumerable: true,
  get: function get() {
    return _colorRanges.COLOR_RANGES;
  }
});
Object.defineProperty(exports, "DefaultColorRange", {
  enumerable: true,
  get: function get() {
    return _colorRanges.DefaultColorRange;
  }
});

var _defaultSettings = require("./default-settings");

var _customColorRanges = require("./custom-color-ranges");

var _colorRanges = require("./color-ranges");
