// kepler.gl/dist/layers/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Layer", {
  enumerable: true,
  get: function get() {
    return _baseLayer["default"];
  }
});
exports.LayerClasses = exports.KeplerGlLayers = void 0;

var _pointLayer = _interopRequireDefault(require("./point-layer/point-layer"));

var _arcLayer = _interopRequireDefault(require("./arc-layer/arc-layer"));

var _lineLayer = _interopRequireDefault(require("./line-layer/line-layer"));

var _gridLayer = _interopRequireDefault(require("./grid-layer/grid-layer"));

var _hexagonLayer = _interopRequireDefault(require("./hexagon-layer/hexagon-layer"));

var _geojsonLayer = _interopRequireDefault(require("./geojson-layer/geojson-layer"));

var _clusterLayer = _interopRequireDefault(require("./cluster-layer/cluster-layer"));

var _iconLayer = _interopRequireDefault(require("./icon-layer/icon-layer"));

var _heatmapLayer = _interopRequireDefault(require("./heatmap-layer/heatmap-layer"));

var _h3HexagonLayer = _interopRequireDefault(require("./h3-hexagon-layer/h3-hexagon-layer"));

var _scenegraphLayer = _interopRequireDefault(require("./scenegraph-layer/scenegraph-layer"));

var _tripLayer = _interopRequireDefault(require("./trip-layer/trip-layer"));

var _baseLayer = _interopRequireDefault(require("./base-layer"));

// Copyright (c) 2019 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
// base layer
// individual layers
var KeplerGlLayers = {
  PointLayer: _pointLayer["default"],
  ArcLayer: _arcLayer["default"],
  LineLayer: _lineLayer["default"],
  GridLayer: _gridLayer["default"],
  HexagonLayer: _hexagonLayer["default"],
  GeojsonLayer: _geojsonLayer["default"],
  ClusterLayer: _clusterLayer["default"],
  IconLayer: _iconLayer["default"],
  HeatmapLayer: _heatmapLayer["default"],
  H3Layer: _h3HexagonLayer["default"],
  ScenegraphLayer: _scenegraphLayer["default"],
  TripLayer: _tripLayer["default"]
};
exports.KeplerGlLayers = KeplerGlLayers;
var LayerClasses = {
  point: _pointLayer["default"],
  arc: _arcLayer["default"],
  line: _lineLayer["default"],
  grid: _gridLayer["default"],
  hexagon: _hexagonLayer["default"],
  geojson: _geojsonLayer["default"],
  cluster: _clusterLayer["default"],
  icon: _iconLayer["default"],
  heatmap: _heatmapLayer["default"],
  hexagonId: _h3HexagonLayer["default"],
  '3D': _scenegraphLayer["default"],
  trip: _tripLayer["default"]
};
exports.LayerClasses = LayerClasses;
